### ISSUE TITLE
Warning log connection pool is full, discarding connection

### ISSUE THREAD

**Issue by dMopp** [dougiteixeira/proxmoxve#76]
This error originated from a custom integration.

Logger: urllib3.connectionpool
Source: custom_components/proxmoxve/__init__.py:236 
Integration: Proxmox VE (documentation, issues) 
First occurred: 11:24:22 (1 occurrences) 
Last logged: 11:24:22

Connection pool is full, discarding connection: <url>. Connection pool size: 10

**Comment by dougiteixeira** [comment]
This warning log is caused outside of the integration, I'll check if there's a way to deal with it...

**Comment by nima-1102** [comment]
+1

Integration Version: 3.1.0

Core: 2023.11.2
Supervisor: 2023.11.3
Operating System: 11.1
Frontend: 20231030.2

```
This error originated from a custom integration.

Logger: urllib3.connectionpool
Source: custom_components/proxmoxve/coordinator.py:404
Integration: Proxmox VE (documentation, issues)
First occurred: 20:13:33 (1 occurrences)
Last logged: 20:13:33

Connection pool is full, discarding connection: 192.168.1.200. Connection pool size: 10
```

```
This error originated from a custom integration.

Logger: urllib3.connectionpool
Source: custom_components/proxmoxve/coordinator.py:496
Integration: Proxmox VE (documentation, issues)
First occurred: 20:13:33 (1 occurrences)
Last logged: 20:13:33

Connection pool is full, discarding connection: 192.168.1.200. Connection pool size: 10
```

**Comment by dmshimself** [comment]
same for me with HA 2023.12.2

**Comment by alexruffell** [comment]
What does this error mean, and is there anything we can do on our end to eliminate or reduce its frequency? I have 3 nodes so that may be part of why I see this warning a lot in my logs

**Comment by ChristophCaina** [comment]
is there any update on this?
I just installed the integration today - and my log is pretty full of those warnings... (at least, three in a row)...

In addition to that, I have an error logged about some time outs... ?

```
Dieser Fehler wurde von einer benutzerdefinierten Integration verursacht

Logger: custom_components.proxmoxve
Source: helpers/update_coordinator.py:306
Integration: Proxmox VE (documentation, issues)
First occurred: 11:08:54 (7 occurrences)
Last logged: 11:08:54

Timeout fetching proxmox_coordinator_disk_homeserver_/dev/sdc data
Timeout fetching proxmox_coordinator_lxc_122 data
Timeout fetching proxmox_coordinator_disk_homeserver_/dev/sdg data
Timeout fetching proxmox_coordinator_disk_homeserver_/dev/sda data
Timeout fetching proxmox_coordinator_lxc_120 data
```

**Comment by dougiteixeira** [comment]
I believe you can filter these logs using the code below in your `configuration.yaml` file:
```
logger:
  default: critical
  filters:
    urllib3.connectionpool:
      - "Connection pool is full"
```

**Comment by DDomnick** [comment]
So am I to understand the solution to this issue of dropping connections is to hide the error?  Isn't there a way to increase the connection pool size >10?

Other report on closed issue that went stale: [https://github.com/home-assistant/core/issues/68487](url)

```
Logger: urllib3.connectionpool
Source: custom_components/proxmoxve/api.py:82
integration: Proxmox VE ([documentation](https://github.com/dougiteixeira/proxmoxve), [issues](https://github.com/dougiteixeira/proxmoxve/issues))
First occurred: 9:27:39 AM (4 occurrences)
Last logged: 9:48:39 AM
Connection pool is full, discarding connection: 10.10.0.121. Connection pool size: 10
```

**Comment by dMopp** [comment]
Hiding is not a solution… Is this a ha core issue or an addon issue? Proxmox addon has a very slow starting time and tons of requests. Maybe optimizing the connections (open, use, close) can solve that, too

**Comment by dougiteixeira** [comment]
The suggestion to hide the logs is just a stopgap solution, the definitive solution depends on a deeper analysis (this same problem occurs with some main Home Assistant integrations), at the moment I am not available to delve deeper into this. Feel free to investigate and propose a definitive solution to the problem.

**Comment by DDomnick** [comment]
Thank you for the update, additional information, and certainly your work on this excellent integration.   I didn't realize this issue was systemic.  I'll take a deeper dive, and will update if i find a more permanent solution.

________________________________

**Comment by ejpenney** [comment]
If I understand correctly, the underlying problem is the connection pool isn't large enough.  You can see [here](https://github.com/dougiteixeira/proxmoxve/commit/e095660b9b7f2cec8cc560e3e48cfcb37a1158f9) a brute-force solution to this problem.  I just did this on my development server, usually I get connection pool messages from Proxmox approximately every 90 seconds after boot, the dev server is getting zero messages for 30 minutes and running.

Two things become immediately apparent:
1. I have no idea what the correct number of connections to have in the pool is, is this dependent on our setup?  I've been looking at this for all-of ten minutes so maybe the answer is obvious to someone else. There may be a static number, or it may need to be something dependent on a specific Home Assistant setup (EG: number of guests we're monitoring?)
2. This appears to be a problem in [proxmoxer](https://github.com/proxmoxer/proxmoxer), the upstream library, I had to patch a private object in this library to get this hack working.  A "proper" fix, would be to either have proxmoxer set the pool size appropriately, or allow this integration to pass a desired pool size into the constructor.

Anyways, I hope this helps someone who knows a bit more about these specific projects...

**Comment by jsapede** [comment]
@ejpenney POC did the trick ! 
will have some lont time test to see 
but wherever i look, proxmoxer or urllib3 they only say to hide the warnings :(

**Comment by umrath** [comment]
I am interested in a solution as well.

**Comment by Trblz42** [comment]
Following this topic - same issue here

`2025-01-27 12:06:14.811 WARNING (SyncWorker_0) [homeassistant.loader] We found a custom integration proxmoxve which has not been tested by Home Assistant. This component might cause stability problems, be sure to disable it if you experience issues with Home Assistant
2025-01-28 22:10:22.799 WARNING (SyncWorker_5) [urllib3.connectionpool] Connection pool is full, discarding connection: 192.168.10.202. Connection pool size: 10`

**Comment by Boss-man** [comment]
@ejpenney is the POC still valid? Tried to look at the current code, but not quite sure where to insert the POC code.

Or maybe better switch to other solution, like glances ?

**Comment by ejpenney** [comment]
> [@ejpenney](https://github.com/ejpenney) is the POC still valid? Tried to look at the current code, but not quite sure where to insert the POC code.
> 
> Or maybe better switch to other solution, like glances ?

Yes, the POC is still valid, at least I'm still running it in my instance with no troubles. The code goes in basically the same places, one import at the top of the file and the two lines at the end of build_client()

I rebased it [here ](https://github.com/dougiteixeira/proxmoxve/commit/0405c63823ed4a57cd6e3185199b220fd3c3ab2d) if you need a more visual reference.

**Comment by kellerassel007** [comment]
Have the same log error

**Comment by dannypv05261** [comment]
> For what its worth, I was able to patch in session caching and reuse to avoid filling the urllib3 connection pool with duplicate sessions. I also upped the pool max size to 50. So far it is working fine on my HA server. If anyone wants to give it a try or finish implementing this and open a PR here's a [link to my fork](https://github.com/tannerln7/proxmoxve). You can add my fork as a custom repo in hacs and install it that way.

I started using your fork two days ago. I can see no warning message prompt anymore.

**Comment by wiedwo** [comment]
tannerln7 and dannypv05261 seem to have found a solution to the problem, why isn't it adopted? 
What are the reasons against it? 
Please comment.

**Comment by ThiloAm** [comment]
I can reproduce this issue with the current integration version 4.0.2.

Environment:
- Home Assistant Core: 2026.5.1
- Proxmox VE integration: 4.0.2
- proxmoxer: 2.2.0
- Proxmox VE: 9.1.6
- 22 Proxmox devices / 387 Home Assistant entities

The warning occurred every 60 seconds, usually 2–4 times in parallel:

```text
WARNING [urllib3.connectionpool]
Connection pool is full, discarding connection: <host>
Connection pool size: 10
```

The integration creates separate coordinators for nodes, updates, disks, ZFS pools, QEMU VMs, LXC containers and storages. These coordinators refresh at the same interval and share the same ProxmoxAPI client, which can result in more than 10 concurrent HTTPS connections.

I tested increasing the underlying requests connection pool in `api.py`:

```python
from requests.adapters import HTTPAdapter

adapter = HTTPAdapter(
    pool_connections=32,
    pool_maxsize=32,
    max_retries=0,
    pool_block=False,
)

session = self._proxmox._store["session"]
session.mount("https://", adapter)
```

After restarting Home Assistant:
- no further `Connection pool is full` warnings
- Proxmox entities continue updating normally
- no regressions observed so far

This appears to be a viable fix, although using the private `_store` attribute may not be the ideal long-term implementation. A supported way to configure the proxmoxer requests session or to stagger coordinator refreshes would likely be preferable.

If there is interest, I can also prepare a pull request with a proper implementation.
